Data inspector tool

KrystianKedron opened the pull request:
The first implementation of data inspector (task s4).
This PR implement the new approach to data inspector functionality similar like in the hightcharts framework. See demo: https://www.highcharts.com/demo/line-ajax.

Commit:
First implementation of the DataInspectorModel which is a general class use to point-picking (S4 task)

Commit:
Implement DataInspectorTool use to:
     - attach the datainspectormodel to the plot
     - connect the mouse move event with proper slot in the datainspectormodel object
     - control enable/disable datainspector mode

Commit:
Attach dateinspector to the trend menu

Commit:
Change the name of the method and correct doc string

Commit:
Fix error with bad position of scatter point symbol (set custom scatter point symbol/size after enable datainspector mode)

Commit:
Fix error with restarting datainspector

Commit:
PEP8 - correct line length

Commit:
Change in DataInspectorTool to inherit from BaseConfigurableClass

Commit:
Create new variables (global date_format and temp parent) to achieve correct line length

cpascual commented on taurus_pyqtgraph/datainspectortool.py:
```diff
@@ -0,0 +1,62 @@
+        self.enable = False
+        self.data_inspector = None
+
+    def attachToPlotItem(self, plot_item, y2=None):
```
Unless I am missing something, I think that the y2 parameter is not used at all in this tool, so it should not be required (unless it is planned to be used later on, but I think it won't be necessary)

KrystianKedron replied:
Yes you are right. I correct this.

cpascual commented on taurus_pyqtgraph/datainspectortool.py:
```diff
@@ -0,0 +1,62 @@
+        menu.addAction(self)
+        self.Y2Axis = y2
+
+    def _onTriggered(self):
```
Very minor thing: maybe this method should be renamed to `_onToggled`?

KrystianKedron replied:
I think a lot about name for this method -> onToggled is good idea.

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,205 @@
+        # Update itemas to make sure the older point is not still trigger
+        curve.updateItems()
+        scatterPoint = self.scatter[i].points()[point_index[0]]
+        scatterPoint.setSize(20)
```
instead of hardcoding, make this `20` a class attribute (so that it can be changed)

KrystianKedron replied:
Should be provides the getter and setter for this variables?

cpascual replied:
I would not bother for now. We can la add them later on if we need them. (the idea is to only create the minimum necessary public API, so that we do not need to maintain it afterwards)

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,205 @@
+                            # create the custom style of the point
+                            # if point style (symbol and size) not exist
+                            curve.setSymbol('o')
+                            curve.setSymbolSize(10)
```
instead of hardcoding, make this `10` a class attribute (so that it can be changed)

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,205 @@
+                            # if point style (symbol and size) not exist
+                            curve.setSymbol('o')
+                            curve.setSymbolSize(10)
+                            self.custom_point_style_init = True
```
this flag (`self.custom_point_style_init`) should be stored *per curve*. Otherwise, if a plot contains 2 curves, one having symbols and the other not having them, the symbols will be erased for **all** the curves when calling `reset_custom_style`

KrystianKedron replied:
Ok, I update this.

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,205 @@
+from taurustrendset import TaurusTrendSet
+
+
+class DataInspectorModel(object):
```
This comment is just an open question since it may be a matter of taste:

In its constructor, `DataInspectorModel` is implicitly "attaching" to the plot item (and effecting changes on the plot item). Likewise, it is implicitly calling `detach` in its destructor. 

I **personally** prefer to be explicit in situations like this. For example, I prefer to explicitly call `detach` instead of simply deleting the object in the `DataInspectorTool._onTriggered` method. The reasons are:
- follow Python Zen of ["explicit better than implicit"](https://www.python.org/dev/peps/pep-0020/)
- It gives the user of the class more control (e.g. the tool could instantiate the inspector just once and attach/detach it instead of instantiating/deleting it each time it is toggled)

KrystianKedron replied:
I update the code and now is more clear what happened. So this is the next reason. 
I like this approach. 
I hope this is what you mean.

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,205 @@
+        - :class:`pyqtgraph.TextItem` - to showing the x,y value
+    Also this class blink the selected scatter point item.
+    """
+    def __init__(self, plot):
```
This is a general comment: as a matter of coding style, we try to keep the **public** API to a minimum in order to facilitate refactorings later on. In parcatice this means that "by default" we make most members (attributes and methods) "protected" by sufixing them with `_` and only leave as "public" those which we explicitly want the users to access.
The philosophy is that it is easier to convert something protected into public if eventually needed than to have to go through a deprecation cycle and leave backwards-compatibility hacks if we want to refactor.

In practice this means that in this class members such as `self.scatter`, `self.informationClouds`, `self.custom_point_style_init`, `self.v_line`, `paintCloud()`, `triggerPoint()`, **etc.** would probably be better as "protected"

KrystianKedron replied:
Ok, this is good point.
I think this is good place to ask this question, but I have problem:
How to properly call the method and variables?
Because Taurus use camelCase for named of the variables. 
Python PEP8 say that for this purpose should be use snake_case.
I use snace_case for named of the variables in the DataInspectorModel (I've noticed it now, but this is my habit) - should i change it to the camelCase? To keep consistency in the whole project? 
The same issue is with the named of the method.

cpascual replied:
the rule in taurus (which IMHO should be followed for taurus_pyqtgraph as well) is to make an exception to PEP8 case recommendation for consistency **in the Qt-related modules**. Any module that is not Qt-related, should follow the general PEP8 case rules. In the case of taurus_pyqtgraph, I think most of it will be Qt-related.

Note that taurus itself is pretty inconsistent since this rule was decided much later than the start of the project, and we decided not to refactor everything. But at least for new code we *try* to follow it

KrystianKedron replied:
That explains a lot, 
Now I see the point.

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,205 @@
+
+            # Iteration over each curves in the plot and checking
+            # that mouse point if is near the x axis date
+            for i, curve in enumerate(filter(lambda o: not isinstance(o, TaurusTrendSet),
```
why filtering out curves that are not trendsets?. IMHO this tool should work for any curve being displayed in the plot item.
Note that the idea is that the tools that we provide should be as generic (and taurus-independent) as possible.

KrystianKedron replied:
Now the sel.plot.curves don't contains only curves object. On the end of this collection is taurustrendset object.

I change this condition to be more agnostic and not creating the taurus dependency. Now I check that the object in this collection is the curve(plotdataitem object).

What you think?

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,205 @@
+            self.plot.scene().removeItem(cloud)
+
+    @staticmethod
+    def getPointIndex(np_curve_data, mouse_point):
```
This method should accept a parameter to set the tolerance (i.e. to be passed to `point_picker.checkPoint` as `roi`). The optimal value of that parameter should be calculated by `DataInspectorModel` taking into account the current scale. More specifically, IMHO the tolerance should be calculated considering the symbol size (e.g. 20 pixels) and translating it to a range in x axis coordinates)

KrystianKedron replied:
This is a good idea to optimize the algorytm finding the point on the curves.

I implemented very simple ROI calculation based on the pixel size and symbol size
(calculation relies on the minimum and maximum value in the X axis).

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,205 @@
+                                           "<span>x=%s "
+                                           "<span>y=%0.1f</span> "
+                                           "</div>" % (self.cloudStyle,
+                                                       self.timestampToTime(x),
```
Converting to time should only be done if the x axis associated to the curve is a `DateAxisItem`

KrystianKedron replied:
Ok, I implement this issue like you say, but now is dependecy to the custom AxisItem 
(DataAxisItem implemented by taurus pyqtgraph - so this isn't very agnostic?)

Commit:
Remove ussles Y2Axis variable

Commit:
Change named of the _onTriggerd method to _onToggled

Commit:
Add new variable to store the trigger point size and default point size (for curve that don't have point size setup!)

Commit:
Fix: now self._cb.toggled connect to proper method

Commit:
Fix: remove y2 variables from attachToPlotItem when TaurusTrend initializing the data inspector object

Commit:
Now each curve have information if the custom_style was setup (to don't reset style for all curves)

Commit:
Create  implicit attaching to the plot and removing from the plot the data_inspector object

Commit:
Create public API

Commit:
Use camelCase to named of the all method in the DataInspectorModel class

Commit:
Change iteration filtering (over curves) to be more agnostic

Commit:
Converting to time x value only if the x axis is the DateAxisItem object

Commit:
Add calculation of roi -> considering the symbol size and x axis range

Commit:
Update doc string's in DataInspectorModel class

Commit:
Correct typo

Commit:
Remove destructor -> raise exception after closing the app

Commit:
Better roi scale in the PointPicker class (prevent raise the maximum recursion exception)

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,270 @@
+
+            # Iteration over each curves in the plot and checking
+            # that mouse point if is near the x axis date
+            for i, curve in enumerate(filter(lambda o: isinstance(o, PlotDataItem),
```
Is this filter really necessary? Maybe just do it for all items in self.plot.curves...

KrystianKedron replied:
Yes, this feature is not necessary, 
but I only want to be sure that I iterate only over curves object, 
because like I said in this collection there are also other objects. 
I tested it without filtering and it worked fine, so I removed not necessary filtering, like you said.

Commit:
Remove filtering in datainspectormodel

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,270 @@
+    def __new__(cls, *args, **kwargs):
+        return np.array(*args, **kwargs).view(PointPicker)
+
+    def checkPoint(self, point_to_check, roi):
```
I think this is essentially the same idea already implemented in `numpy.isclose`

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,270 @@
+    This class extend the :class:`numpy.ndarray` to handle the numpy array
+    with curve data.
+    """
+    def __new__(cls, *args, **kwargs):
```
Just a note on style. Maybe this is a justified use of `__new__`, but IMHO in most cases one can (should) use `__init__` instead for writing higher-level more pythonic code. See https://stackoverflow.com/a/674369

cpascual commented on taurus_pyqtgraph/datainspectormodel.py:
```diff
@@ -0,0 +1,270 @@
+
+                # Getting the nearest point in the x axis
+                # (if exist - empty array if not)
+                point_index = self._getPointIndex(curve_x_data, mousePoint.x())
```
I think that this implementation could be simplified if we used the approach proposed [here](https://groups.google.com/forum/#!msg/pyqtgraph/HU4GsfzAxXU/H8UPYbPjAwAJ). 

I created [a very basic example](https://gist.github.com/cpascual/5620f913f83cae4be2a4f50057111e4c) which follows that approach and with which we would get rid of the `PointPicker`, `_calculateRoi` and `_getPointIndex`

KrystianKedron replied:
This implementation is great!
I modify the datainspectormodel like in the example and did small changes:

- To prevent shaking the graph on an automatic scale in` _createLabels` I use `setParentItem` method on the `text_item`. In this case pyqtgraph doesn't take into account the text_item and not adjust the scale (this couse the shaking the graph when we pick the point near the border of the plot).

- I add the point trigger logic, and also if needed converting the x value to the time.
Also I fromat the y value to a defined number of decimal places (like in y_format variable).

Commit:
Update the datainpectormodel to use new logic to finding the nearest point (like in cpascual InspectorLine example).

Commit:
Update the datainspectortool to use new method from datainspectormodel

Commit:
Merge branch 'master' into pointpicking

Fixed conflicts in:
	taurus_pyqtgraph/trend.py

Commit:
Add copyright header

Commit:
Rename DataInspectorModel --> DataInspectorLine

Commit:
Fix indentation issue

The changed lines were run inside a for loop, but they should be done
just once at the end

Commit:
Refactor the inspector tool

- simplify the logic and do some variable renames
- make the inspector line usable without the inspector tool
- add highlight markers instead of depending on existing symbols

Commit:
Auto-close context menu when enabling inspector tool

Auto-close the menu so that the user can start inspecting. Only do it
when enabling.

Commit:
(m) Add TODO

cpascual commented:
I did some refactoring and updated with the latest master.
IMHO this is ready to be integrated

Commit:
Add the data inspector tool to TaurusPlot

Commit:
Move all inspector implementation to datainspectortool.py

Merge datainspectorline.py into datainspectortool.py and do some
further code and docstring cleaning.
Also add a taurus-less example in the main section.

Commit:
Register the checked status of the inspector as a config property

Commit:
Merge branch 'master' into pointpicking

cpascual commented:
Idea for improvement: Instead of showing x,y in each label, the x value could be an extra tag at the base (or top) of the vertical line

cpascual commented:
The current implementation seems enough as a first step, hence I am merging it. However it should be improved with (at least) y2-axis support
